Add FromOpenPMDPulse profile

fabidie opened the pull request:
The FromOpenPMDPulse profile is an IncidentField method that reads the laser pulse from an INSIGHT measurement (at the focus position) into PIConGPU.

Since the INSIGHT measurement is done in the frequency domain, the E-field data first has to be transformed into the time domain. Afterwards, the field data chunk is saved to openPMD  and from there it will be read onto the incident field plane at the start of the simulation volume (via linear interpolation). 
Those "preparation steps" are done with a python script, which is provided under picongpu/lib/python/picongpu/extra/input. There, one can find also an example workflow which explains the preparation routine more in detail.

Currently, there are some restrictions in using this profile:
- propagation and polarization only work along the axes, **not** diagonally!
- the pulse will always be aligned centered at the incident plane
- one needs to increase the reserved GPU memory size in memory.param - since the field data chunk will be stored on each used device at timestep 0 (and not during init, before allocating the particle memory), the simulation could otherwise run into memory issues.
- the Insight measurement scales have to be mm and fs, otherwise the preparation routine has to be adapted

There have been various tests to validate this new laser profile:
#### 1. Using a gaussian pulse profile

A Gaussian pulse field chunk ($\lambda$ = 800nm, $\tau$ = 30fs, $w_0$ = 10 $\mu$ m, $z$ = $-z_R$ = -0.393mm) has been read from OpenPMD into the simulation volume and then propagated to the focus. First, the amplitude at the incident plane ("PIC") has been compared to the analytical definition of the Gaussian pulse at $z$ = $-z_R$ ("Original") for different time steps covering the pulse duration.
![incidentPlaneIte600](https://github.com/user-attachments/assets/3b2efe4e-8134-4cea-809e-68c3a2632e1a)
The vertical line in the last subplot indicates the position of the incident plane at this moment. The amplitudes look similar in extent, but different in structure and amplitude, since the longitudinal sampling of both datasets is different. These small differences in space/time cause the visible differences, since the amplitude varies with $\lambda$ in the longitudinal direction.

Then, the amplitude at the focal plane ("PIC") has been compared to the analytical definition of the Gaussian pulse at the focus ("Original"), also for different time steps covering the pulse duration.
![focusPlaneIte5666](https://github.com/user-attachments/assets/8afac928-bd7d-4410-80b1-66d918bd746c)
(amplitude in the focal plane)
![focusDiffIte5666](https://github.com/user-attachments/assets/9fdfe405-4b28-4de6-9f61-95efee15c9d9)
(transversely integrated amplitude, the black line represents the focus position)
![xzPlaneIte5666](https://github.com/user-attachments/assets/dc1a2af8-3969-4561-902d-9ab6310af198)
(amplitude in propagation direction, the blue lines represent the focus position)
Here, the lines again represent the focal plane position for this certain time step. Both pulses look similar in extent and amplitude.

Furthermore, the beam energy (density) was investigated. The original input beam had an energy of 4.5 J and the measured beam energy in the focus  (shown in the pictures above) was a bit below with 4.3 J. The result could probably be improved with a finer sampling. Also, the energy density in the focus (integrated along the propagation axis)  looks similar in extent and value (and is probably a better indicator than the beam amplitudes at incident and focal plane, since the effects of different spacing vanish due to the integration):
![energyDensityIte5666](https://github.com/user-attachments/assets/5db9a96d-9206-4030-a4e6-ada3f7dfd4ce)
The energy was calculated according to $W = \int dV \frac{\epsilon_0}{2} (E^2 + c^2 B^2)$ and the energy density $\mathcal{W} = \int dz \frac{\epsilon_0}{2} (E^2 + c^2 B^2)$.

Finally, the simulation has been re-run with the GaussianPulse profile (and the same parameters) to compare it to the former results (they looked the same).

#### 2. Using a Gauss-Laguerre pulse
According to the definition of a Gauss-Laguerre pulse in the GaussianPulse profile, a Gauss-laguerre pulse field chunk has been written to OpenPMD and read into PIConGPU. The pulse parameters were the same as for the Gaussian pulse above, just $\tau$ was set shorter to $30/2\sqrt{2 ln 2}$ fs. As Laguerre coefficients, the example set in GaussianPulse.def was used. For two different positions, the simulation results ("PIC") were compared to the analytical solution ("Original"):
**a) at $z \approx -z_R$**
![xyPlaneIte1285](https://github.com/user-attachments/assets/3cb76852-3538-4070-a02c-d07335f26317)
(amplitude along the longitudinal direction)
![energyDensityIte1285](https://github.com/user-attachments/assets/7bff1c78-9354-4fe9-b588-757baa913a8e)
(energy density, definition see above)
**b) at the focus position**
![xyPlaneIte5597](https://github.com/user-attachments/assets/9998e582-3699-4eb6-bea8-8808bdd07aea)
(amplitude along the longitudinal direction,  the blue line is the focus position at this time step)
![energyDensityIte5597](https://github.com/user-attachments/assets/86ed0411-5105-45cb-b127-03055f89b350)
(energy density, definition see above)
![focusDiffIte5597](https://github.com/user-attachments/assets/9d821f14-f151-4883-8aec-0283ffe7b495)
(transversal integrated amplitude,  the black line is the focus position at this time step)
In the above mentioned simulations, the transversal simulation cell was square-shaped, as well as the transversal extent of the test pulse field chunk. To test also the asymmetrical case, the simulation cell in $x$ direction was halved in size. The results were similar: 
![energyDensityIte5649](https://github.com/user-attachments/assets/da0c589e-66de-408e-86b5-e98837e78668)
(energy density for an asymmetrical cell size; the asymmetry in amplitude shows that the energy is conserved better for a smaller cell size)
Then, the cell size was reset to original (transversal square-shaped), but the (original) field chunk was chosen to be asymmetrical:
![originalChunkExtent](https://github.com/user-attachments/assets/da0dc1ad-8faa-4856-a804-f716f1691225)
(original Gauss-Laguerre field at $z=-z_R$, before reading it into the simulation)
The resulting simulation did behave just the same as above, e.g. the pulse was still centered and not distorted after propagating it into the focus:
![yxPlaneIte5597](https://github.com/user-attachments/assets/cb033851-697a-491d-91fa-53fa3cb3808d)
![yzPlaneIte5597](https://github.com/user-attachments/assets/ba755a14-0abe-4561-a0d0-d4db99054e1f)
(amplitude in longitudinal direction, along both transversal directions; blue is the focus position)
![energyDensityIte5597](https://github.com/user-attachments/assets/dd2179ae-cc30-4d07-ac46-45cc9687d300)
( energy density)

#### 3. Using an Insight measurement

For that, the measured field data has been propagated out of the focus (-150, -300, -500 and -1280 (= $-z_R$ ) $\mu$ m, during the preparation routine), saved to OpenPMD and then read into the simulation. This has been done for different propagation and polarization directions, for different (transversal) simulation window extents, for different transversal grid spacing ($w_0/6$ and $w_0/10$) and for a different number of devices (1, 2, 4).
The field at the incident plane ("PIC") has been compared to the read-in ("Original") data:
![incidentPlaneIte1800](https://github.com/user-attachments/assets/62b5dc94-1d23-4b16-a61d-995fedb08e99)
(propagation in z direction, polarization in x direction, just as the original)
![incidentPlaneNegPolIte1800](https://github.com/user-attachments/assets/aede2351-2ee6-4574-b64a-4c3b5b8616b9)
(propagation in z direction, polarization in -x direction; this automatically also swaps the y axis to keep the original pulse orientation)
![xzPlaneNegPropIte1800](https://github.com/user-attachments/assets/eadcde87-654a-4e7d-8e1e-dd20ca3df3d9)
(propagation in -z direction, polarization in x direction; the time axis of the original data has been transformed to a spatial one by multiplication with $c$ to allow a better comparison)

For z = -300 $\mu$ m, the pulse has been propagated in PIConGPU into the focus and then compared there to the Insight measurement at the focus: 
![focusDiffIte4400](https://github.com/user-attachments/assets/aa999148-0a8c-4493-8ac7-fa487f08932c)
(amplitude at the beam center in propagation direction)
![focusPlaneIte4400](https://github.com/user-attachments/assets/3a3e6bb9-53b9-4187-aa47-d2cd9c607d78)
(amplitude in focal plane)
![xzPlaneIte4500](https://github.com/user-attachments/assets/d1616822-43be-4687-ab22-c2b4db00748f)
(amplitude in longitudinal direction, blue is the focal plane; the time axis of the original data has been transformed to a spatial one by multiplication with $c$ to allow a better comparison)

Differences in the above pictures arise from:
- different spacing (so they are not completely at the same place in spacetime)
- using the $c t$ approximation for the longitudinal space axis
- numerical errors ($N_\lambda = 10$ as longitudinal grid spacing, could be done more fine)
- slightly different results when propagating the pulse with the angular spectrum method (as in the preparation routine) compared to a Maxwell solver (here, Yee has been used)

As another test, the beam energy has been calculated. The Original Insight data pulse had an energy of 4.5 J and the energy of the same pulse in PIConGPU was about 2 - 3% less (using the Yee solver), which could be improved by a finer sampling.

finnolec committed:
Add openPMD output for transitionRadiation plugin.

finnolec committed:
Make transitionRadiation plugin a multi-plugin

finnolec committed:
Further clean-up for transition radiation plugin

* Make text-based output optional
* Make foilPosition a command-line parameter
* Clang-format
* Make TransitionRadiation plugin dependent on openPMD
* Update CTR example

finnolec committed:
Update openPMD call and rename foilPositionY

* Optimize memory footprint of openPMD call
* Rename foilPosition to foilPositionY
* Remove foilPosition from param file
* Update foilPosition in example

finnolec committed:
Update transition radiation plugin docs

SergeyKonstantinErmakov committed:
Maxwell-Juettneri distributions for highly relativistic particle temperatures

finnolec committed:
Implement @PrometheusPi's suggestions

fabidie committed:
Added InsightPulse profile

fabidie committed:
Added InsightPulse profile

BrianMarre commented:
@fabidie Our CI is unhappy with the code formatting, you ned to run the code formatting tools to fix most and fix the remaining code style issues.

Please install the pre-commit code formater hooks as described [here](https://picongpu.readthedocs.io/en/latest/dev/docs/COMMIT.html) and run ``pre-commit run --all-files`` in your repository afterwards.

fabidie committed:
fixed code formatting

fabidie committed:
fixing pmacc include backets

fabidie committed:
fixed dimensionality and type conversions

fabidie committed:
took care of openPMD API dependency

fabidie committed:
only allow 3D simulations

steindev commented on include/picongpu/fields/incidentField/Traits.hpp:
```diff
@@ -171,6 +183,16 @@ namespace picongpu
 
+#if(ENABLE_OPENPMD == 1) && (SIMDIM == DIM3)
+            //! Specialization for InsightPulse profile which has unknown amplitude
+            // kann man die Templates so leer lassen und es fkt trd?
```
Outdated question?

fabidie replied:
Oh yes, I checked every file for stupid comments except for Traits haha. I'll remove it

steindev commented on include/picongpu/fields/incidentField/profiles/InsightPulse.def:
```diff
@@ -0,0 +1,122 @@
+
+                        /** Datatype of the field record
+                         *
+                         * OpenPMD needs this information at compile time.
```
openPMD is written with lowercase 'o'

fabidie replied:
I changed it, hopefully I caught every uppercase O 😅

steindev commented on include/picongpu/fields/incidentField/profiles/InsightPulse.def:
```diff
@@ -0,0 +1,122 @@
+                        static constexpr float_64 DIRECTION_Y = 1.0;
+                        static constexpr float_64 DIRECTION_Z = 0.0;
+
+                        /** Unit E polarization direction in PIConGPU
```
So far, this seems to be the only time where polarisation is written with 'z'

fabidie replied:
In BaseParam.hpp it is spelled "polarization" as well. Since I used some methods from there, I also copied the 'z'; but I will change it to 's' for consistency reasons.

steindev commented on include/picongpu/fields/incidentField/profiles/InsightPulse.def:
```diff
@@ -0,0 +1,122 @@
+                {
+                    struct InsightPulseParam
+                    {
+                        /** Path to the openPMD input file
```
I think you should mention here, how this openPMD file can be obtained from the measurement file.
It might be best to describe this workflow in the manual, e.g. as a new section in `/usage/workflows/addLaser.html` and just refer to the documentation here.

fabidie replied:
Okay, I'll do that👍🏼

steindev commented on include/picongpu/fields/incidentField/profiles/InsightPulse.hpp:
```diff
@@ -0,0 +1,618 @@
+
+                    /** Singleton to load field data from OpenPMD to device
+                     *
+                     * The complete dataset will be loaded (equally) to all GPUs, aswell as
```
-> 'as well as'

fabidie replied:
changed ✅

steindev commented on include/picongpu/fields/incidentField/profiles/InsightPulse.hpp:
```diff
@@ -0,0 +1,618 @@
+                     *
+                     * Right now, the data will be loaded at timestep 0, which means that the user
+                     * has to **increase the reserved GPU memory** in memory.param, since otherwise
+                     * the simulation will run into memory issues.
```
This is an important prerequisite which should also be mentioned in the documentation and the example Param struct.

fabidie replied:
Done ✅

steindev commented on include/picongpu/fields/incidentField/profiles/InsightPulse.hpp:
```diff
@@ -0,0 +1,618 @@
+                     * since this would solve this bug.
+                     *
+                     * @tparam T_Params user parameters, providing filename etc.
+                     */
```
Actually, you can also add a todo list here or at the beginning of the file with all the known limitations that should be tackled in refactorings.

fabidie replied:
Done ✅

steindev commented on include/picongpu/fields/incidentField/profiles/InsightPulse.hpp:
```diff
@@ -0,0 +1,618 @@
+                            bufferExtentOpenPMD = std::make_shared<pmacc::HostDeviceBuffer<float_X, 1u>>(3u);
+                            bufferCellSizeOpenPMD = std::make_shared<pmacc::HostDeviceBuffer<float_X, 1u>>(3u);
+                            bufferOffsetOpenPMD = std::make_shared<pmacc::HostDeviceBuffer<float_X, 1u>>(3u);
+
```
You assign to three different variables the same thing. Why do you have three variables at all?

fabidie replied:
I'm not so familiar with the concept of shared pointers in C++. If there is another way to solve this, please let me know - the fact is that I need 3 float-vectors of length 3 to describe the field data chunk completely (extent, cell size and the offset w.r.t. the simulation window). I calculate those values at the same timestep when the data is loadad  and since I need them outside the openPMD-singleton, I put them in a HostDeviceBuffer (just like the corresponding field data chunk).

steindev commented on include/picongpu/fields/incidentField/profiles/InsightPulse.hpp:
```diff
@@ -0,0 +1,618 @@
+                             * data at the chunk borders will be discarded. The maximum discarded value (relative to
+                             * the maximum amplitude) will be logged. A possible improvement of this code could include
+                             * storing just the necessary parts of the recorded field data.
+                             */
```
Do you decide only on the basis of cell numbers? But the cell sizes in the data file and PIConGPU can be different, can't they? Shouldn't you decide on the basis of real space sizes?

fabidie replied:
Yes, thats true, and thats why I converted the openPMD cell size into internal units (so they have the same length scale as `cellSize`)
https://github.com/fabidie/picongpu/blob/9acaccdb80cd952359076fff86189de4af313195/include/picongpu/fields/incidentField/profiles/InsightPulse.hpp#L263C33-L264C110
```C++
dataBoxCellSize(aligningAxisIndex[d])
     = static_cast<float_X>(cellSizeRaw[d] * mesh.gridUnitSI()) / UNIT_LENGTH;
```

steindev commented on include/picongpu/fields/incidentField/profiles/InsightPulse.hpp:
```diff
@@ -0,0 +1,618 @@
+                            } // d
+
+                            // This message is shown for every used device, but 1x would be sufficient (aswell as the
+                            // loop before)
```
Regarding the message, you can put an if statement with the rank number which evaluates to true only for rank 0.

fabidie replied:
Thanks!

steindev commented on include/picongpu/fields/incidentField/profiles/InsightPulse.hpp:
```diff
@@ -0,0 +1,618 @@
+                        }
+
+                    private:
+                        OpenPMDdata()
```
Is it important to declare this constructor as `HINLINE`? @psychocoderHPC

steindev commented on include/picongpu/fields/incidentField/profiles/InsightPulse.hpp:
```diff
@@ -0,0 +1,618 @@
+                                       and (posPIC[d] < offsetOpenPMDdataBox(d) or posPIC[d] > offsetOpenPMDdataBox(d) + (extentOpenPMDdataBox(d) - 1.0_X) * cellSizeOpenPMDdataBox(d))
+                                   or (timeOriginPIC - Unitless::TIME_DELAY) < 0.0_X
+                                   or (timeOriginPIC - Unitless::TIME_DELAY) > (extentOpenPMDdataBox(d) - 1.0_X)
+                                           * cellSizeOpenPMDdataBox(d) / SPEED_OF_LIGHT)
```
I would be grateful if you could indicate with another pair of parenthesis with which block of `or` conditions the `and` condition is evaluated.

fabidie replied:
I have reworked this whole section (better structure and more comments) to make it better understandable. This also solves your next two comments.

steindev commented on include/picongpu/fields/incidentField/profiles/InsightPulse.hpp:
```diff
@@ -0,0 +1,618 @@
+                                    else // keep original direction
+                                        idxClosestRaw[d]
+                                            = (posPIC[d] - offsetOpenPMDdataBox(d)) / cellSizeOpenPMDdataBox(d);
+                                }
```
same as above.

steindev commented on include/picongpu/fields/incidentField/profiles/InsightPulse.hpp:
```diff
@@ -0,0 +1,618 @@
+                                }
+                            }
+
+                            // correct longitudinal index
```
What does correct mean here? In the sense of 'fix' or 'choose the right'

fabidie replied:
I meant it in the sense of 'fix' since the for-loop in the lines before just tackles the transversal index, but produces nonsense for the longitudinal one. But (as written two commetns above) I changed this line anyway so that I do not have to 'fix' anything at all.

steindev commented on include/picongpu/fields/incidentField/profiles/InsightPulse.hpp:
```diff
@@ -0,0 +1,618 @@
+                         *
+                         * @param totalCellIdx cell index in the total domain (including all moving window slides)
+                         */
+                        HDINLINE float_X getValueE(floatD_X const& totalCellIdx) const
```
This routine is very long. Can you split it a little in subroutines, e.g. outsourcing the index finding and interpolation?

fabidie replied:
Done ✅

steindev reviewed (commented):
Good work so far! There are a couple of questions/comments/wishes from my side, of course :grin:

I did not review the python code so far. Will do this next week.

fabidie committed:
reviewer comments

fabidie committed:
reviewer comments

steindev commented on docs/source/usage/workflows/addLaser.rst:
```diff
@@ -46,14 +46,20 @@ The differences between matching incident field- and laser profiles are:
-For a laser profile with non zero field amplitudes on the transversal borders of the profile e.g. defined by the profile ``Free`` without a transversal envelop the trait ``MakePeriodicTransversalHuygensSurfaceContiguous`` must be specialized and returning true to handle field periodic boundaries correctly.
+For a laser profile with non zero field amplitudes on the transversal borders of the profile e.g. defined by the profile ``Free`` without a transversal envelope the trait ``MakePeriodicTransversalHuygensSurfaceContiguous`` must be specialized and returning true to handle field periodic boundaries correctly.
+
+Another special case is the ``InsightPulse`` profile, which reads an experimental Laser pulse into the simulation, which was obtained from an Insight measurement.
```
How about: "which reads a laser pulse profile, defined in frequency-space domain, from an openPMD file into the simulation. Specifically this allows to read a pulse profile obtained from an (Insight) measurement"?

Coming to think of it: Maybe we should change the profile's name to reflect its universal applicability. I guess something along the lines of `PulseFromFile` (similar to the density profile) would fit?

fabidie replied:
I changed it to `FromOpenPMDPulse`

steindev commented on docs/source/usage/workflows/addLaser.rst:
```diff
@@ -46,14 +46,20 @@ The differences between matching incident field- and laser profiles are:
+For a laser profile with non zero field amplitudes on the transversal borders of the profile e.g. defined by the profile ``Free`` without a transversal envelope the trait ``MakePeriodicTransversalHuygensSurfaceContiguous`` must be specialized and returning true to handle field periodic boundaries correctly.
+
+Another special case is the ``InsightPulse`` profile, which reads an experimental Laser pulse into the simulation, which was obtained from an Insight measurement.
+It is less flexible in choosing propagation and polarisation direction, since those are only allowed along the cell edges, not diagonally.
```
-> "For now, it is ..."

steindev commented on include/picongpu/fields/incidentField/profiles/InsightPulse.def:
```diff
@@ -0,0 +1,131 @@
+                         * data, propagate it to the desired position and transform it into the time domain before
+                         * saving the resulting E-field chunk to openPMD.
+                         *
+                         * This field chunk wil be stored at timestep 0 of the simulation on every used device, which
```
-> "will"

steindev commented on lib/python/picongpu/extra/input/preparingInsightData.py:
```diff
@@ -0,0 +1,472 @@
+Copyright 2024 Fabia Dietrich
+"""
+
+import openpmd_api as io
```
I think we wanted to change `io` due to possible ambiguity. Please use `opmd` instead. See #4802

steindev commented on lib/python/picongpu/extra/input/preparingInsightData.py:
```diff
@@ -0,0 +1,472 @@
+
+def supergauss2D(xy, A, x0, y0, w, n=4):
+    """
+    2D supergaussian distribuition
```
-> "distribution"

steindev commented on lib/python/picongpu/extra/input/preparingInsightData.py:
```diff
@@ -0,0 +1,472 @@
+    2D gaussian distribution
+
+    Arguments:
+    xy: coordinates
```
Since you do ` x,y = xy` later, you should specify more specifically what kind of data structure (and its extents) is expected here.

steindev commented on lib/python/picongpu/extra/input/preparingInsightData.py:
```diff
@@ -0,0 +1,472 @@
+    2D supergaussian distribuition
+
+    Arguments:
+    xy: coordinates
```
same as above

steindev commented on lib/python/picongpu/extra/input/preparingInsightData.py:
```diff
@@ -0,0 +1,472 @@
+    return g.ravel()
+
+
+class preproutines:
```
Class names are typically written camelcase: "PrepRoutines"

steindev commented on lib/python/picongpu/extra/input/preparingInsightData.py:
```diff
@@ -0,0 +1,472 @@
+c = 2.99792e-4  # speed of light in mm/fs
+
+
+def gauss2D(xy, A, x0, y0, w):
```
please do `A` -> `amp` to comply with code style (lower case for variables and meaningful name)

steindev commented on lib/python/picongpu/extra/input/preparingInsightData.py:
```diff
@@ -0,0 +1,472 @@
+    """
+    class to manipulate and prepare Insight data for PIConGPU usage:
+        - phase corrections
+        - artefact corrections in the near field
```
-> "artifact"

steindev commented on lib/python/picongpu/extra/input/preparingInsightData.py:
```diff
@@ -0,0 +1,472 @@
+
+    def __init__(self, path, name, foc):
+        """
+        Load Insight data (= far field), propagate to near field
```
please write "Insight data (in focus = far field)", so this is clearly defined once.

steindev replied:
You should also mention that the field is assumed to be the first (=0th) group in the h5 file

steindev commented on lib/python/picongpu/extra/input/preparingInsightData.py:
```diff
@@ -0,0 +1,472 @@
+        """
+        Load Insight data (= far field), propagate to near field
+        and extract (gaussian) fit parameters.
+        The far field has to be measured in dependence of "x", "y" and "w".
```
-> "in dependence of the transverse coordinates "x", "y", and frequency "w"."

steindev commented on lib/python/picongpu/extra/input/preparingInsightData.py:
```diff
@@ -0,0 +1,472 @@
+        self.zR = self.waist**2 / (2 * c) * self.w[self.wc_idx]
+
+        # propagate to near field (right before the lens, i.e. d=0)
+        preproutines.FFtoNF(self)
```
should be `self.FFtoNF()`. Well, actually it should be `self.ff_to_nf()` in order to comply with our standard code style. i.e. function names are all lowercase with underscores.

steindev commented on lib/python/picongpu/extra/input/preparingInsightData.py:
```diff
@@ -0,0 +1,472 @@
+        """
+        Since there is no information about the global phase of every wavelength measured with Insight,
+        there has to be a phase correction.
+        For that, we assume perfect compression of the beam center in the near field, thus we substract
```
-> 'subtract'

steindev commented on lib/python/picongpu/extra/input/preparingInsightData.py:
```diff
@@ -0,0 +1,472 @@
+        Since there is no information about the global phase of every wavelength measured with Insight,
+        there has to be a phase correction.
+        For that, we assume perfect compression of the beam center in the near field, thus we substract
+        a global phase for every wavelength.
```
", thus we subtract a frequency's phase value at the beam center from every phase value in the frequency's  transverse distribution in the NF." is more precise and correct, isn't it?

steindev commented on lib/python/picongpu/extra/input/preparingInsightData.py:
```diff
@@ -0,0 +1,472 @@
+        Furthermore, there is the possibility to add dispersion parameters (GVD, TOD) to the data.
+
+        Arguments:
+        GVD : Group velocity disperion in fs**2/rad,  optional. The default is 0.
```
-> "dispersion"

(good to see that you do not copy-paste :wink:)

steindev commented on lib/python/picongpu/extra/input/preparingInsightData.py:
```diff
@@ -0,0 +1,472 @@
+
+        self.Ew_NF = np.abs(self.Ew_NF) * np.exp(
+            1j * (np.angle(self.Ew_NF) - phase_centr + dispersion)
+        )  # correcting near field
```
please put the comment above the line. Same for the next line.

steindev commented on lib/python/picongpu/extra/input/preparingInsightData.py:
```diff
@@ -0,0 +1,472 @@
+        yidx_low = y_ugly_idx - uglybins
+        yidx_upp = y_ugly_idx + uglybins
+
+        x_surr = np.concatenate((self.x_NF[xidx_low - 3 : xidx_low], self.x_NF[x_ugly_idx + uglybins : xidx_upp + 3]))
```
This magic `3` is another safety margin? Would you add a comment explaining that number please? Also, the rule of thumb 'Whatever you use twice, automate!' applies here too...

steindev replied:
And why do you apply it only along the x direction? (Comment in code please)

steindev commented on lib/python/picongpu/extra/input/preparingInsightData.py:
```diff
@@ -0,0 +1,472 @@
+        self.Ew_NF[yidx_low:yidx_upp, xidx_low:xidx_upp, :] = interp((Y_hole, X_hole, W_hole))
+
+        # correct the far field
+        preproutines.NFtoFF(self)
```
-> `self.nf_to_ff()`

steindev commented on lib/python/picongpu/extra/input/preparingInsightData.py:
```diff
@@ -0,0 +1,472 @@
+        print("new near field center coordinates: xc = %.2f mm, yc = %.2f mm" % (self.xc_NF, self.yc_NF))
+
+        # correct the far field
+        preproutines.NFtoFF(self)
```
-> `self.nf_to_ff()`

steindev commented on lib/python/picongpu/extra/input/preparingInsightData.py:
```diff
@@ -0,0 +1,472 @@
+        If the near field is not centered around (0, 0), it can be done with this function.
+        It automatically also corrects the far field by propagating the centered near field back into the focus.
+        """
+        nx = int(self.xc_NF / self.dx_NF + 0.5)  # number of pixels to shift in x-direction
```
comment above the line please. Same for the next line.

steindev commented on lib/python/picongpu/extra/input/preparingInsightData.py:
```diff
@@ -0,0 +1,472 @@
+
+        if nx == 0 and ny == 0:
+            print("already centered, no corrections necessary")
+            return 0
```
You should probably check whether `nx` and `ny` are within half the size of the data extent. Otherwise you just shift the pulse from one half of the screen to the other, don't you?

steindev reviewed (commented):
Further comment. Reviewed until `propagate()` method in `lib/python/picongpu/extra/input/preparingInsightData.py`.

steindev commented on lib/python/picongpu/extra/input/preparingInsightData.py:
```diff
@@ -0,0 +1,472 @@
+        w_exp = self.waist * np.sqrt(1 + (z / self.zR) ** 2)
+        assert w_exp < 0.2 * min(
+            self.dx * self.Ew.shape[0], self.dy * self.Ew.shape[1]
+        ), "Oops, you wanted to propagate too far!"
```
please add "The pulse will not fit in the transverse window."

steindev commented on lib/python/picongpu/extra/input/preparingInsightData.py:
```diff
@@ -0,0 +1,472 @@
+        a = np.fft.fftfreq(self.x.size, d=self.dx)
+        b = np.fft.fftfreq(self.y.size, d=self.dy)
+        B, A, W = np.meshgrid(b, a, self.w)
+        cond = (2 * np.pi * c / W) ** 2 * (A**2 + B**2)
```
Would you please explain in a comment in the code what this mask is for? (or maybe reference to an equation in Godman if appropriate)

steindev commented on lib/python/picongpu/extra/input/preparingInsightData.py:
```diff
@@ -0,0 +1,472 @@
+        print("far field propagated to z = %.2f mm" % (z))
+        return np.fft.ifft2(F_Ew, axes=(0, 1)) * lin_phase
+
+    def toTimeDomain(self, Ew, lamb_supp=10):
```
please rename to `to_time_domain()`

steindev commented on lib/python/picongpu/extra/input/preparingInsightData.py:
```diff
@@ -0,0 +1,472 @@
+            (self.y, self.x, self.w), Ew, bounds_error=False, fill_value=0, method="linear"
+        )
+        Ew_fft = fft_interp((Y_fft, X_fft, W_fft))
+        # set the negative frequency part to 0 (instead of the complex conjugate) and neglegt the imaginary part afterwards
```
-> "neglect"

steindev commented on lib/python/picongpu/extra/input/preparingInsightData.py:
```diff
@@ -0,0 +1,472 @@
+
+        print("field data size: %.1f GB" % (np.prod(self.Et.shape) * 8 * 10**-9))  # assume datatype double
+
+    def saveToOpenPMD(self, outputpath, outputname, energy, pol="x", crop_x=0, crop_y=0, crop_t=0):
```
please rename to `save_to_openpmd()`

steindev commented on lib/python/picongpu/extra/input/preparingInsightData.py:
```diff
@@ -0,0 +1,472 @@
+
+    def saveToOpenPMD(self, outputpath, outputname, energy, pol="x", crop_x=0, crop_y=0, crop_t=0):
+        """
+        Save the field data in time domain to an OpenPMD checkpoint. This output will be used for the InsightPulse profile.
```
-> "openPMD"
-> "FromOpenPMDPulse" profile

steindev reviewed (commented):
reviewed until `saveToOpenPMD`

BrianMarre committed:
link compile tests as examples of FLYonPIC use

steindev commented on lib/python/picongpu/extra/input/preparingInsightData.py:
```diff
@@ -0,0 +1,472 @@
+    def saveToOpenPMD(self, outputpath, outputname, energy, pol="x", crop_x=0, crop_y=0, crop_t=0):
+        """
+        Save the field data in time domain to an OpenPMD checkpoint. This output will be used for the InsightPulse profile.
+        The propagation axis ("z") will be transformed via z = c*t to a spatial one.
```
-> "The pulse's time evolution at a specific z position will be transformed to a spatial evolution along the propagation direction via z=c*t".

Well, actually this procedure is certainly wrong to do, as we often encounter pulses with length of the order of the Rayleigh length such that the true spatial evolution is affected by defocusing.

In general, it is possible to store several iterations in one openPMD file, which is what we should do here. But this might complicate reading the file and the initialization procedure in PIConGPU. I cannot estimate the effort to perform this refactoring and since you back transform from z to t when reading in PIConGPU, such that the whole workflow is consistent, I guess it is ok to leave it as is for the moment.

However, you should put a big fat Attention Note that this procedure is in principle not correct and needs to be refactored. Please put such a note at the beginning of this file and in the beginning of PIConGPU's FromOpenPMDPulse file.

fabidie replied:
That is true. This transformation step is nonsense from a physical poin of view, it is just done to fulfil the openPMD requirements to store a field chunk (in just one iteration) and undone stright away again by the FromOpenPMDPulse profile. Certainly it would be more correct to store each time slice in their respective iteration, but since there are around 1000 of them per chunk, it will not only complicate the source code but probably also increase the computation time (I guess, didn't try it out yet). 
Nevertheless, I added it to the refactoring list.

steindev commented on lib/python/picongpu/extra/input/preparingInsightData_example.ipynb:
Is there a specific reason to leave this as a jupyther notebook and not check it in as a python script? I ask because notebooks require more maintenance than scripts...

And a another point: Please add the copyright notice.

fabidie replied:
Because it looks nice 😅 
With the plots and comment sections in between all the necessary preparaion steps I can easily explain what happens and why it has to happen. But I can put it in a python script anyway, if you think that is better.

steindev commented on lib/python/picongpu/extra/input/preparingInsightData_example.ipynb:
```diff
@@ -0,0 +1,467 @@
+    "3. propagate to the near field\n",
+    "4. fit the near field data with a 2D supergaussian to extract the beam center and waist size\n",
+    "\n",
+    "For that, we provide path, filename and focal distance (same unit as the transversal scales) to the init funcion."
```
Change from 'we' to 'you need to' in order to be consistent with the beginning of the file.

steindev replied:
typo "function"

steindev commented on lib/python/picongpu/extra/input/preparingInsightData_example.ipynb:
```diff
@@ -0,0 +1,467 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# some nice colorful pictures of our data\n",
```
our -> your

steindev commented on lib/python/picongpu/extra/input/preparingInsightData_example.ipynb:
```diff
@@ -0,0 +1,467 @@
+    "ax1.set_title(\"spectral integrated amplitude, far field\")\n",
+    "ax1.set_xlabel(\"x [mm]\")\n",
+    "ax1.set_ylabel(\"y [mm]\")\n",
+    "ax1.set_xlim(-0.2, 0.2)\n",
```
Better delete this manually set limits? They might have a surprising effect.

steindev commented on lib/python/picongpu/extra/input/preparingInsightData_example.ipynb:
```diff
@@ -0,0 +1,467 @@
+    "    origin=\"lower\",\n",
+    "    extent=(insight.x[0], insight.x[-1], insight.y[0], insight.y[-1]),\n",
+    ")\n",
+    "ax1.set_title(\"spectral integrated amplitude, far field\")\n",
```
-> "spectrally"?

steindev commented on lib/python/picongpu/extra/input/preparingInsightData_example.ipynb:
```diff
@@ -0,0 +1,467 @@
+    "    origin=\"lower\",\n",
+    "    extent=(insight.x_NF[0], insight.x_NF[-1], insight.y_NF[0], insight.y_NF[-1]),\n",
+    ")\n",
+    "ax2.set_title(\"spectral integrated amplitude, near field\")\n",
```
-> "spectrally"?

steindev commented on lib/python/picongpu/extra/input/preparingInsightData_example.ipynb:
```diff
@@ -0,0 +1,467 @@
+    "ax2.set_title(\"spectral integrated amplitude, near field\")\n",
+    "ax2.set_xlabel(\"x [mm]\")\n",
+    "ax2.set_ylabel(\"y [mm]\")\n",
+    "ax2.set_xlim(-100, 100)\n",
```
Delete limits?

steindev commented on lib/python/picongpu/extra/input/preparingInsightData_example.ipynb:
```diff
@@ -0,0 +1,467 @@
+    ")\n",
+    "ax3.set_title(\"phase in near field beam center\")\n",
+    "ax3.set_xlabel(r\"$\\omega$ [rad/fs]\")\n",
+    "ax3.set_ylabel(\"rad/fs\")\n",
```
the phase unit is probably just 'rad', isn't it?

steindev commented on lib/python/picongpu/extra/input/preparingInsightData_example.ipynb:
```diff
@@ -0,0 +1,467 @@
+    "ax4.set_title(r\"SD$_x$ in focus\")\n",
+    "ax4.set_ylabel(\"x [mm]\")\n",
+    "ax4.set_xlabel(r\"$\\omega$ [rad/fs]\")\n",
+    "ax4.set_ylim(-0.2, 0.2)\n",
```
Delete limits?

steindev commented on lib/python/picongpu/extra/input/preparingInsightData_example.ipynb:
```diff
@@ -0,0 +1,467 @@
+    "ax5.set_title(r\"SD$_y$ in focus\")\n",
+    "ax5.set_xlabel(r\"$\\omega$ [rad/fs]\")\n",
+    "ax5.set_ylabel(\"y [mm]\")\n",
+    "ax5.set_ylim(-0.2, 0.2)\n",
```
Delete limits?

steindev commented on lib/python/picongpu/extra/input/preparingInsightData_example.ipynb:
```diff
@@ -0,0 +1,467 @@
+    "        np.sum(\n",
+    "            np.abs(\n",
+    "                insight.Ew[\n",
+    "                    np.abs(insight.y - insight.yc + 2 * insight.waist).argmin() : np.abs(\n",
```
I cannot follow this calculation, you need to explain it offline :sweat_smile:

steindev commented on lib/python/picongpu/extra/input/preparingInsightData_example.ipynb:
```diff
@@ -0,0 +1,467 @@
+   "source": [
+    "## Correct the data\n",
+    "### Adjust beam compression and add dispersion parameters\n",
+    "Before we go on with any calculations, we have to correct the phase. Insight reconstructs the amplitude of the far field beam aswell as the phase, up to a unknown global phase for every frequency.\n",
```
there is "we" again...

steindev replied:
-> "as well as"

steindev replied:
-> "an unknown"

steindev commented on lib/python/picongpu/extra/input/preparingInsightData_example.ipynb:
```diff
@@ -0,0 +1,467 @@
+   "metadata": {},
+   "source": [
+    "### Correct ugly spots in the near field (optional)\n",
+    "Sometimes, the amplitude in the near field looks weird, showing some (unphysical) peaks or holes. These can cause artefacts in the far field and will thus be smoothened out."
```
-> "artifacts"

steindev commented on lib/python/picongpu/extra/input/preparingInsightData_example.ipynb:
```diff
@@ -0,0 +1,467 @@
+    "    origin=\"lower\",\n",
+    "    extent=(insight.x[0], insight.x[-1], insight.y[0], insight.y[-1]),\n",
+    ")\n",
+    "ax1.set_title(\"spectral integrated amplitude, far field\")\n",
```
-> "spectrally"

steindev commented on lib/python/picongpu/extra/input/preparingInsightData_example.ipynb:
```diff
@@ -0,0 +1,467 @@
+    "    origin=\"lower\",\n",
+    "    extent=(insight.x_NF[0], insight.x_NF[-1], insight.y_NF[0], insight.y_NF[-1]),\n",
+    ")\n",
+    "ax2.set_title(\"spectral integrated amplitude, near field\")\n",
```
-> "spectrally"

steindev commented on lib/python/picongpu/extra/input/preparingInsightData_example.ipynb:
```diff
@@ -0,0 +1,467 @@
+    "    origin=\"lower\",\n",
+    "    extent=(insight.x[0], insight.x[-1], insight.y[0], insight.y[-1]),\n",
+    ")\n",
+    "ax1.set_title(\"spectral integrated amplitude, far field\")\n",
```
-> "spectrally"

steindev commented on lib/python/picongpu/extra/input/preparingInsightData_example.ipynb:
```diff
@@ -0,0 +1,467 @@
+    "    origin=\"lower\",\n",
+    "    extent=(insight.x_NF[0], insight.x_NF[-1], insight.y_NF[0], insight.y_NF[-1]),\n",
+    ")\n",
+    "ax2.set_title(\"spectral integrated amplitude, near field\")\n",
```
-> "spectrally"

steindev commented on lib/python/picongpu/extra/input/preparingInsightData_example.ipynb:
```diff
@@ -0,0 +1,467 @@
+    "ax1.set_title(\"spectral integrated amplitude, far field\")\n",
+    "ax1.set_xlabel(\"x [mm]\")\n",
+    "ax1.set_ylabel(\"y [mm]\")\n",
+    "ax1.set_xlim(-0.2, 0.2)\n",
```
delete limits?

steindev commented on lib/python/picongpu/extra/input/preparingInsightData_example.ipynb:
```diff
@@ -0,0 +1,467 @@
+    "ax2.set_title(\"spectral integrated amplitude, near field\")\n",
+    "ax2.set_xlabel(\"x [mm]\")\n",
+    "ax2.set_ylabel(\"y [mm]\")\n",
+    "ax2.set_xlim(-100, 100)\n",
```
delete limits?

steindev commented on lib/python/picongpu/extra/input/preparingInsightData_example.ipynb:
```diff
@@ -0,0 +1,467 @@
+    "ax1.set_title(\"spectral integrated amplitude, far field\")\n",
+    "ax1.set_xlabel(\"x [mm]\")\n",
+    "ax1.set_ylabel(\"y [mm]\")\n",
+    "ax1.set_xlim(-0.2, 0.2)\n",
```
delete limits?

steindev commented on lib/python/picongpu/extra/input/preparingInsightData_example.ipynb:
```diff
@@ -0,0 +1,467 @@
+    "ax2.set_title(\"spectral integrated amplitude, near field\")\n",
+    "ax2.set_xlabel(\"x [mm]\")\n",
+    "ax2.set_ylabel(\"y [mm]\")\n",
+    "ax2.set_xlim(-100, 100)\n",
```
delete limits?

steindev commented on lib/python/picongpu/extra/input/preparingInsightData_example.ipynb:
```diff
@@ -0,0 +1,467 @@
+    "plt.title(\"spectral integrated amplitude, propagated to %.2f mm\" % (z))\n",
+    "plt.xlabel(\"x [mm]\")\n",
+    "plt.ylabel(\"y [mm]\")\n",
+    "# plt.xlim(-.2, .2)\n",
```
delete limits?

steindev commented on lib/python/picongpu/extra/input/preparingInsightData_example.ipynb:
```diff
@@ -0,0 +1,467 @@
+   "source": [
+    "## Propagate\n",
+    "Now the far field data is ready to be propagated. For that, the angular spectrum method is used. \n",
+    "Watch out not to propagate too far, since then the growing beam diameter could reach the transversal window borders and thus cause fourier transform artefacts."
```
-> "artifacts"

psychocoderHPC committed:
fix wrongly used HDINLINE and missing inline

- `inline` is added to function where it was missed and there disallow
  using multiple compile units
- change wrongly used `HDINLINE`
- add missing `HDINLINE`

steindev commented on lib/python/picongpu/extra/input/preparingInsightData_example.ipynb:
```diff
@@ -0,0 +1,467 @@
+    "    extent=(insight.t[0], insight.t[-1], insight.x[0], insight.x[-1]),\n",
+    ")\n",
+    "ax1.set_ylabel(\"x [mm]\")\n",
+    "ax1.set_xlim(-100, 100)\n",
```
delete limits?

steindev commented on lib/python/picongpu/extra/input/preparingInsightData_example.ipynb:
```diff
@@ -0,0 +1,467 @@
+    ")\n",
+    "ax2.set_xlabel(\"t [fs]\")\n",
+    "ax2.set_ylabel(\"y [mm]\")\n",
+    "ax2.set_xlim(-100, 100)\n",
```
delete limits?

steindev commented on lib/python/picongpu/extra/input/preparingInsightData_example.ipynb:
```diff
@@ -0,0 +1,467 @@
+   "id": "8b5e94b1-d37e-4639-ad5c-8990206df22e",
+   "metadata": {},
+   "source": [
+    "## Save data to OpenPMD\n",
```
-> "openPMD"

steindev commented on lib/python/picongpu/extra/input/preparingInsightData.py:
Small but valuable feature suggestion: Since you have the frequency domain field you, you can easily provide the field envelope in time domain (i.e. get rid of the field oscillations), which eases pulse analysis tremendously. You only need to calculate `self.Et.real**2+self.Et.imag**2`.

Or you can show this as an additional example plot in the jupyter notebook...

fabidie replied:
I added this information to the notebook. 
In the corresponding field plots I already worked with the envelope instead of the (oscillating) field (since `self.Et` is already complex), and now I added another plot, showing both field and envelope to make it more clear. I also calculated an estimation for the pulse duration (measured in the beam center).

steindev reviewed (commented):
Now I am through all files once. I did not look yet at your recent changes after the first part of the review.

Overall, you did a very good job @fabidie!
...only small things I have seen and commented. And a few requests :sweat_smile:

fabidie committed:
renaming pulse profile; reviewers comments on python part

fabidie committed:
renaming pulse profile; reviewers comments on python part

psychocoderHPC committed:
add `PluginRegistry`

Add helper construct to register atomatically plugin at the start of the
application.

psychocoderHPC committed:
register each plugin via registry

Register plugins at the start of the application.

psychocoderHPC committed:
remove `components.param` and `starter.param`

- remove `components.param` and `starter.param`
- move PIConGPU starter configuration into `main.cpp`

BrianMarre committed:
reviewer suggestions

ikbuibui committed:
Bin particles leaving the simulation

psychocoderHPC committed:
remove conda from our docuemantion and scripts

Conda is using as default software channel the anaconda reposetory.
Software in this channel is not always free to use and can require a
license. Therefore we do not point the user to this possible problem.

psychocoderHPC committed:
add header validation

Cheks if all PIConGPU header files include there dependencies directly or
transitiv.

fabidie committed:
fix asymmetric field chunk bug

fabidie committed:
add optional aperture in mid field

psychocoderHPC commented:
@fabidie I rebased our PR and pushed it again to fix CI issues

psychocoderHPC committed:
refactor ID provider

- remove global device variable to hold the ID provider state
- ID provider does not depend on the simulation dimensionallity anymore
- create a default ID provider named `globalId`

It is now possible to have multiple ID provider.

psychocoderHPC committed:
PMacc: refactor particle implementation

- add assign operator to the particle
- add method `copyAndInit()`

psychocoderHPC committed:
add support for id generator to compile time interfaces

psychocoderHPC committed:
add support for id generator to value_identifier

psychocoderHPC committed:
use id generator andrefactored interfaces

fabidie committed:
add disperion measurement methods

fabidie committed:
fixing AD measurement

ikbuibui committed:
Fixes

ikbuibui committed:
mark arguments as maybe unused

psychocoderHPC committed:
move field code into cpp files

Move parts of the field code into own compile units.
A new file `difines.hpp` is required to solve param dependencies without
an update of all param files.

psychocoderHPC committed:
fix nvcc bug

nvcc is creating constructors the class `KernelMetaData` and complains
about destructing the string from a devíce function.
Even declaring all constructors is not solving the nvcc warning.

```
/pmacc/lockstep/Kernel.hpp(60): warning #20011-D: calling a __host__ function("std::__cxx11::basic_string<char,     ::std::char_traits<char> , ::std::allocator<char> > ::~basic_string()") from a __host__ __device__ function("pmacc::exec::detail::KernelMetaData::~KernelMetaData") is not allowed

Remark: The warnings can be suppressed with "-diag-suppress <warning-number>"

pmacc/lockstep/Kernel.hpp(60): warning #20011-D: calling a __host__ function("std::__cxx11::basic_string<char,     ::std::char_traits<char> , ::std::allocator<char> > ::basic_string(const ::std::__cxx11::basic_string<char,     ::std::char_traits<char> , ::std::allocator<char> > &)") from a __host__ __device__ function("pmacc::exec::detail::KernelMetaData::KernelMetaData") is not allowed
```

psychocoderHPC committed:
compile unit: FieldJ and stage compute current

psychocoderHPC committed:
fix missing include

psychocoderHPC committed:
compile unit: stage refactoring

Separate collision, atomic phy., iteration start and particle init.

psychocoderHPC committed:
compile unit: stage refactoring

Separate collision, atomic phy., iteration start and particle init.

psychocoderHPC committed:
compile unit: stage refactoring

Separate collision, atomic phy., iteration start and particle init.

psychocoderHPC committed:
compile unit: synchrotron radiation

ikbuibui committed:
update JURECA profile and tpl
SBATCH --distribution=*:block is no longer recommended
MPI-settings/CUDA needs to be loaded for CUDA aware MPI to work properly

BrianMarre committed:
add explict check for atomicPhyiscs electron species

psychocoderHPC committed:
compile unit: IO plugins, FIeldTmp, collision and ionization

ikbuibui committed:
Add profiles for JEDI@JSC

psychocoderHPC committed:
compile unit: transform plugins

Particle binning, binner, charge conservation, radiation, transition
radiation, calorimeter, phasespace and marcro particles per supercell

psychocoderHPC committed:
compile unit: refactor last plugins

CountParticles, Emittance, EnergyFields, EnergyParticles, SumCurrents

psychocoderHPC committed:
fix collision compile issues

BrianMarre committed:
add missing space to PlantUMl documentation

psychocoderHPC committed:
unit test: radiaton window and form factor tests

Add test to compare host and defice resuts for both functor types.

psychocoderHPC committed:
rename compileParticlePusher to compile

This test will cover diffeent compile variations to avoid to many
independent tests to reduce the CI test time.

psychocoderHPC committed:
merge compile tests to reduce CI load

psychocoderHPC committed:
rename cmopile test

psychocoderHPC committed:
Reduce number of compile tests

Equally to #5052 the number of compile tests is reduced by merging two
tests. This test is not part of `tests/compile` because particle
operations tested within this compile test can be compile time
expensive.

psychocoderHPC committed:
iCI: reduce number of pusher test combindations

All combinations of pusher are compiled in `compileParticlePusher`
therefore the number of compile tests can be reduced for CI runs.

finnolec committed:
Fix polarization component in shadowgraphy plugin

fabidie committed:
fixing beam energy calculation for applied aperure

BrianMarre committed:
fix Documentation

removes old comment no longer accurate

psychocoderHPC committed:
PMAcc: refactor `math::Vector`

Refactor `math::Vector`
  - add constexpr to nearly all operations
  - default storage policy will be `std::array`
  - remove navigator policy
`compile-time/Vector`
  - refactor runtime vector creation

psychocoderHPC committed:
add basic unit tests for vector

BrianMarre committed:
remove superflous cast

psychocoderHPC committed:
fix possible deadlock

PMacc's event system must be empty before global blocking methods e.g.
MPI collectives are called. OpenPMD's flush is possibly executing
blcoking collectives.

psychocoderHPC committed:
refactor param file structure

- move `TYPICAL_NUM_PARTICLES_PER_MACROPARTICLE` from particles into
  picongpu namespace
- add missing includes

psychocoderHPC committed:
update docuemntation

point to `Simulation-param` where we pointed to `grid.param` before

psychocoderHPC committed:
refactor include order

- remove grid.param
- add new file `simulation.param`
- move `BASE_DENSITY_SI` to simulation.param
- move `TYPICAL_PARTICLES_PER_CELL` to simulation.param
- reduce cyclic includes by adding more files to `defines.hpp`

BrianMarre committed:
rename TransitionDataSet to ChooseTransitionGroup

psychocoderHPC committed:
vector unit test: add nvcc 11.6 workaround

psychocoderHPC committed:
rename grid.param to simulation.param

BrianMarre committed:
update out of source documentation

psychocoderHPC committed:
introduce global object to querry simulation parameters

INtroduce an global object to replace accessing parameters from the
param files directly.

psychocoderHPC committed:
rename variable name `sim` to `simHeader`

Avoid conflicts with the global variable `sim`.

Commit:
remove using cellSize and CELL_VOLUME

BrianMarre committed:
refactor atomicPhysics todos

Commit:
`CELL_HEIGHT` to `sim.pic.getCellSize().y()`

Commit:
`CELL_DEPTH` to `sim.pic.getCellSize().z()`

Commit:
rename `DELTA_T` to `sim.pic.getDt()`

Commit:
`UNIT_TIME` to `sim.unit.time()`

Commit:
`UNIT`LENGHT` to `sim.unit.length()`

Commit:
`TYPICAL_NUM_PARTICLES_PER_MACROPARTICLE` to `sim.unit.typicalNumParticlesPerMacroParticle()`

Commit:
`UNIT_MASS` to `sim.unit.mass()`

Commit:
`UNIT_CHARGE` to `sim.unit.charge()`

Commit:
`UNIT_ENERGY` to `sim.unit.energy()`

Commit:
`UNIT_EFIELD` to `sim.unit.eField()`

Commit:
`UNIT_BFIELD` to `sim.unit.bField()`

psychocoderHPC committed:
fix TWTS renamings

psychocoderHPC committed:
remove `unit.param`

psychocoderHPC committed:
move description of units from deleted files to `sim`

Commit:
`CELL_WIDTH` to `sim.pic.getCellSize().x()`

BrianMarre committed:
reviewer comments

BrianMarre committed:
switch to static cast short hands

psychocoderHPC committed:
fix picongpu unit test

psychocoderHPC committed:
fix synchrotron radiation floating point overflow warning

fix #5036

Fix warning be reorder the terms to stay within the floating point
range.

psychocoderHPC committed:
add review comment changes

psychocoderHPC committed:
reduce CI compile time for atomicPhysics compile test

- avoid compiling derived fields (added fileOutput.param)
- remove unused particle and density functors.

psychocoderHPC committed:
move example SingleParticleTest to tests

Move single particle example to tests and rename it to `SingleParticle`.

psychocoderHPC committed:
CI: PIConGPU reduce number of variations per test

Reduce the number of variations for PIConGPU CI test to one for all exept the
`compile` test where two combinations are compiled to guarantee that we
cover 64bit.

psychocoderHPC committed:
PMacc: fix constexpr math::Vector

- fix CUDA 11.6 compile time evaluation issues by
  - unsing `{}` instead of `()` for type generation
  - constexpr lambda capture by reference instead of copy

psychocoderHPC committed:
CI: reduce number of compile combinations

- `n_wise_generator.py` add option `--quick` to disable compiler chaine
  combination with each example. This will reduc ethe number of tests
with the disadvantage that maybe some test will not compiled with e.g.
nvcc
- activate quick tests for user PRs for examples, tests and benchmarks
   - PRs merged to the dev branch will perfrom all tests

BrianMarre committed:
fix in code comment Counter shape

psychocoderHPC committed:
rename in comments: grid.unitless into simulation.unitless

psychocoderHPC committed:
CI: control via commit message

- add possibility to controll the CI via commit messages
- add basic documention

psychocoderHPC committed:
unit test: fix warning

The bug was already fix with #5066 but removing the attribute completly
is the better way.
```
share/picongpu/unit/MoveParticle.cpp(87): warning #20011-D: calling a __host__ function("bool ::isApproxEqual< ::pmacc::math::Vector<float, (unsigned int)3u,  ::pmacc::math::ArrayStorage<float, (unsigned int)3u> > > (const T1 &, const T1 &)") from a __host__ __device__ function("ParticleStub::operator == const") is not allowed
```

psychocoderHPC commented:
@fabidie please do not forget to rebase the PR to the current development branch

fabidie committed:
renaming internal units

psychocoderHPC committed:
CI:: fix download micro mambda

HZDR is blocking all `anaconda` server. If we download micromamba via
the offical link `https://micro.mamba.pm/api/micromamba/linux-64/latest`
the load balancer is sometimes pointing us to the anaconda mirrow which
is not accesible from within HZDR.

The current workaround is to pull micromamba 1.5.9 directly from github.

ci: no-compile

psychocoderHPC committed:
avoid using SIMDIM for controlflow

fix #5062

Avoid using the macro SIMDIM for program control-flow.
SIMDIM should only be used for include guards.
Fix missing include in untested code IsaacPlugin.x.cpp.

psychocoderHPC committed:
change atomic physics particle attributes

Atomic physics particle attributes was set by default to a possible
valid value but in case an algorithm forget to set the value correctly
we got a false valid value.
By setting it to an invalid value such errors should become visible at
leat by a crash of the simulation.
An additional advantage is that we can now give an electron some of the
properties which allows to track from which ion state and which
transition state the electron was created.

psychocoderHPC committed:
refactor `value_identifier_func`

Refactor `value_identifier_func` and `SetAttributeToDefault` to allow
deriving any kind of attribute from the source particle.

psychocoderHPC committed:
`UNIT_SPEED` to `sim.unit.speed()`

psychocoderHPC committed:
`BASE_CHARGE` to `sim.pic.getBaseCharge()`

psychocoderHPC committed:
`BASE_MASS` to `sim.pic.getBaseMass()`

- removei file `speciesConstants.unitless`

Commit:
`ElectronCharge` to `sim.pic.getElectronCharge()`

Commit:
`ELECTRON_MASS` to `sim.pic.getElectronMass()`

Commit:
`SPEED_OF_LIGHT` to `sim.pic.getSpeedOfLight()`

BrianMarre committed:
add ADK rate calculation

BrianMarre committed:
add enum for ADK laser Polarization

BrianMarre committed:
add field ionization as new group

BrianMarre committed:
update ChooseTransitionGroupKernel

BrianMarre committed:
update FillLocalRateCache

psychocoderHPC committed:
Plugin: move PngWriter to compile unit

create an own compile unit for the PNG plugin.

ci: picongpu

psychocoderHPC committed:
compile unit for pusher stage

psychocoderHPC committed:
move stage particle boundaries to compile unit

ci: picongpu

BrianMarre committed:
update ChooseTransitionKernel

BrianMarre committed:
update todos

psychocoderHPC committed:
cleanup `particlesFunctors.hpp`

To reduce the compile time for independent compile units solver calls were moved to the corresponding stages.

psychocoderHPC committed:
`Simulation.hpp` remove unused includes

ci: picongpu

psychocoderHPC committed:
movev debye length check to own compile unit

ci:: picongpu

psychocoderHPC committed:
fix compile

During the cherry pick a atomatically resolved merge conflict failed.
This commit fixes the issue.

psychocoderHPC committed:
make precisionCast `constexpr`

- add unit test

BrianMarre committed:
implement field Ionization electron init

BrianMarre committed:
remove mentions of old unit systems

BrianMarre committed:
fixes

psychocoderHPC committed:
remove duplicated comments

With PR
https://github.com/ComputationalRadiationPhysics/picongpu/pull/5090 we
missed to remove some comments from the cpp file.

ci: picongpu

psychocoderHPC committed:
PMacc: refactor FileSystem.hpp

- remove dimension from the filesystem class
- remove Filesystem from pmacc::Environment<>

chillenzer committed:
Adjust openPMD alias in python

chillenzer committed:
Add auomatic adjustment docs

BrianMarre committed:
update examples

BrianMarre committed:
seperate atomicDataBoxes checks

ci: picongpu

BrianMarre committed:
add missing getMargin includes to ionizers

BrianMarre committed:
rename ApplyPressureIonization to ApplyIPDIonization

psychocoderHPC committed:
fix include issues

finnolec committed:
Update transition radiation plugin to use sim object

PrometheusPi committed:
move lower case file one directory up

PrometheusPi committed:
adjust include path

done via script:
for f in `grep -l -r absorber.hpp . | grep -v ./.git/`; do ls $f; sed -i 's%/absorber/absorber.hpp%/absorber.hpp%g' $f ; done

chillenzer committed:
Remove instructions from docs

psychocoderHPC committed:
CI: add PIConGPU header check

- rename `test/picHeaderCheck` to `est/picongpuHeaderCheck`
- add ci tests downstream pipe for header checks

ci: picongpu

psychocoderHPC committed:
fix `ci: picongpu` command

Fix ci control, we missed to create an empty job for PMacc pipes.

chillenzer committed:
Add Lenz to zenodo

psychocoderHPC committed:
￼precision selection for PIConGPU quantities

Allow choosing the precision for PIC quantities.
By default, the user-selected precision for PIC quantities.

ci: picongpu

psychocoderHPC committed:
irefactor cell type definitions

- remove the trait getCellType
- remove cell type centered
- create file `ÝeeCell.hpp` and move Yee cell definition to this file
- refactor incident field traits to resolve include order issues

To simplify the code we removed the possibility to use different cell types in PIConGPU.
We kept it in the past but in the current code, we often used already very often the Yee cell and any algorithms e.g. current depeosion require the Yee cell.

PrometheusPi committed:
add method to load distributed amplitudes

PrometheusPi committed:
add documentation on distributed amplitudes in python module

BeyondEspresso committed:
Update frontier-olcf profiles

BeyondEspresso committed:
Bump rocm to version 6.2.0

chillenzer committed:
Clarify use of snakemake params in docs

SimeonEhrig committed:
fix warning in PyPIConGPU version tests

- If the script `requirements_txt_modifier.py` should modify a dependency in a requirements.txt, which does not exist, it throws a warning because this is unexpected behavior. Because warnings are hard to detect in the CI, change it to an error.
- Implement a function to ignore specific environment variables which are set to change the version in a requirements.txt.

BrianMarre committed:
reviewer comments

ci: picongpu

psychocoderHPC committed:
set minimum CMake version back to 3.22

revert #5004

THe feature (link time optimizations) why we increased CMake to 3.25 is
not required anymore because due to other clang-cuda issues with LTO we
changed our implementations.
Supporting older CMake version avoids on some systems that we need our
hand isntalled CMake package.

BrianMarre committed:
rename pressure ionization data to IPDIonization data

psychocoderHPC committed:
CI: update compile matrix

- add tests for
  - HIP 56 - 6.1
  - CUDA 12.2 - 12.5
  - boost 1.79 - 1.85
- update filter rules

ci: full-compile

psychocoderHPC committed:
fix smaller CI issues

- PIConGPU header check
  - fix copy past project name
  - remove define and take PMacc's tpp into account
- fix typo

BrianMarre committed:
fixes

ci: picongpu

BrianMarre committed:
fix error in fixAtomicState Kernel for unintialized ions

ci: picongpu

ikbuibui committed:
Fix cmake prefix path for mpiInfo

psychocoderHPC committed:
refactor fix atomic state kernel

Unify memory access for better performance and fewer code duplications.

BeyondEspresso committed:
Re-add cmake module
ci: no-compile

BrianMarre committed:
update default atomicPhysics_Debug.param comments

updates the documentation of atomicPhysics_debug.param to make limits of
debug print switches clear

BrianMarre committed:
udpate species.param

Commit:
Updating alpha centauri profiles at TUD-ZIH
ci: no-compile

BrianMarre committed:
fix out of bounds memory access in RollForOverSubscription

Commit:
Squashed 'thirdParty/cuda_memtest/' changes from 345628a628..7affd32ac7

7affd32ac7 Merge pull request #50 from psychocoderHPC/fix-clang18Issues
d0508c924e fix clang cuda issues

git-subtree-dir: thirdParty/cuda_memtest
git-subtree-split: 7affd32ac7cd6af7574e36c3a020e416af848218

psychocoderHPC committed:
fix missing include

psychocoderHPC committed:
remove clang extension usage

fix warning

```
/builds/hzdr/crp/picongpu/include/picongpu/../picongpu/plugins/openPMD/WriteSpecies.hpp:424:40: note: read of non-const variable 'mpiSize' is not allowed in a constant expression
/builds/hzdr/crp/picongpu/include/picongpu/../picongpu/plugins/openPMD/WriteSpecies.hpp:337:26: note: declared here
  337 |                 uint64_t mpiSize = gc.getGlobalSize();
      |                          ^
/builds/hzdr/crp/picongpu/include/picongpu/../picongpu/plugins/openPMD/WriteSpecies.hpp:402:46: warning: variable length arrays in C++ are a Clang extension [-Wvla-cxx-extension]
  402 |                     uint64_t allNumParticles[mpiSize];
```

ikbuibui committed:
Update hemeraIn5min.rst

Fix wrong absolute path with relative path

psychocoderHPC committed:
replace EPS0,Z0,MUE0

Replace old variables from `physicalConstants` and use `sim` the
object.

ci: picongpu

psychocoderHPC committed:
fix unit electric field definition

fix bug introduced with #5066

psychocoderHPC committed:
refactor `value_identifier`

This PR provide the possibility that a value_identifier can fully
describe the operation performed for the
  - first initialization -> `initValue()`
  - if a particle attribute is copied -> `copyValue()`
  - if a particle attribute is derived -> `deriveValue()`

For PIConGPU the binary manipulator `Assign` is renamed into `Derive`
and a `Copy` manipulator is added.

With the new description of value identifier there is no need that we
handle special attributes e.g. `particleId` explicit in any algorihm
where we derive particles.

psychocoderHPC committed:
refactor: `HBAR`, `RYDBERG_ENERGY` and `BOHR_RADIOUS`

Move variables to corresponding function into the PIConGPU `sim` object.
Provide methods to convert from joule to eV and back.

ci: picongpu

Commit:
Squashed 'thirdParty/alpaka/' changes from ab3092357b..5c5a690f48

5c5a690f48 add new targets and move to if
df1841403e Implement markdown generator for compiler backend compatibility table (#2388)
11ab218299 Reduce Uniform/Independent-Elements iterator register footprint (#2383)
63ad43fd1e Clean formatting
dad38aaa1f Explicitly call alpaka::detail to achieve SYCL compilation
db5736555d Use the generic address space for SYCL atomic operations (#2389)
3d043fec64 Removed deprecated command Changed from alapka enqueue to alpaka exec Added entries to get - pitches - mdspan - kernelFunctionAttributes - NDIndex
6f759e3737 Fix race condition in image write Improve stability condition Dont validate for guard cells Improve contrast in image write Make changes more obvious
a017c606d1 Babelstream benchmark activation and refactoring
3575cfd8cf Add runtime macro alpaka_throw_acc
fe262f6dc1 Add support for Boost 1.83-1.86
246987e6f7 Add support for CMake 3.27-3.30
c7edaa76bb Add support for ROCm 6.1 and 6.2
6d0bb24fba Add support for CUDA 12.6
36bb18a2f2 Update log messages
5c75f81584 Let the compiler find the HIP headers when building host-only code
2e9b9ae1c7 Add missing includes
70fca93d9f Use gcnArchName instead of gcnArch
20f75cce62 Rewrite the 3D buffer copy example using different uniformElements loops
ca19f519dc Fix uniformElementsAlong<Dim>, etc.
c8d5b79da5 Rename uniformElements[XYZ] to uniformElementsAlong[XYZ], etc.
b4dc73a818 Update to oneAPI 2024.2
2e394b3aca Ignore -Wswitch-default when using clang
53392f4f7a Add 2D heat equation example with shared memory and validation
546f267302 Rewrite vectorAdd using alpaka::uniformElements
f6cd229230 Include iso646 to enable extra keywords on MSVC
05030c08fb Mark the deduction guide for vector as ALPAKA_FN_HOST_ACC (#2376)
6bdf13a00f Refactor the SYCL backend (#2374)
71cbff3f9f fix error in Buffer constructor
a3e562b553 Enable MappedBuf support for the SYCL backend
032e98c2ec ignore -Wswitch-default when using clang (#2373)
1b8710ec57 Fix boolean values after copying from device to host
621e8f9f9d Workaround for a weird bug in oneAPI 2024.x
566b27c3fc Ignore memory leaks in the clang OpenMP library
40a4ff6dcf Add unit tests for IndependentElements kernel loops
76c7e1d131 Add IndependentElements kernel loops
b2b5afa198 Add unit tests for UniformElements kernel loops
e3a9708ec3 Add UniformElements kernel loops
18d26fbf59 Add unit tests for oncePerGrid and oncePerBlock utility functions
1c91bda49b Add oncePerGrid and oncePerBlock utility functions
b0d1157732 Rename `threadsPerGrid` to `elementsPerGrid` (#2362)
9559147719 Rename `threadElemExtents` into `threadElemExtent` (#2363)
be8752c4d5 Add ORCID for Julian Lenz (#2364)
00c51bb399 alpaka CI: set CMAKE_CXX_COMPILER instead using update-alternatives
a25a2fdcfc remove second docker container launch in GitHub Actions Linux Jobs
d7c459d739 remove `KernelBundle`, change signature of `[get|is]ValidWorkDiv*` (#2349)
ad77f15894 fix BufFpgaSyclIntel template arguments (#2361)
d641d8ba08 Add some color output to CI install scripts (#2359)
2ed0f24aea add the possibility to disable exit on error via environment variable in a CI job
cdeae8ad12 fix several bugs in the CI scripts
65afe9ff2f fix `set +euo pipefail` in travis_retry and docker_retry
ac0afb30b4 add colored output to the CI scripts
e4d9e5509b Add missing #include (#2351)
9cf09f11f0 update zenodo.json (#2350)

git-subtree-dir: thirdParty/alpaka
git-subtree-split: 5c5a690f48aeedc6b8651e5b5f486767c12f2be8

psychocoderHPC committed:
refactor AU units

- add conversion au units -> si
- add conversion au unis -> pic

ci: picongpu

psychocoderHPC committed:
change interface `conv.*` to `conv().*

nvcc is not possible to pull static constexpr member into the device
code, therefore acessing a function is required.

ikbuibui committed:
Add new CI Variable PIC_CI_MAINLINE

psychocoderHPC committed:
fix gcc 13 compile issue

Add check to silence gcc 13 array bound check.

ci: full-compile

BrianMarre committed:
add check for valid typical ppc

BrianMarre committed:
update PICMI unsupported error mesage for moving window

psychocoderHPC committed:
apply review comments

- add `automicUnit`
- rename `ev` to `eV`

psychocoderHPC committed:
review changes second round

fabidie committed:
Adding recommendation for time step adjustment

psychocoderHPC committed:
complete include refactoring

- remove `simulation_defines.hpp`
- add `param.hpp` to get access to param files for a subgroup of the
  code
- fix missing includes

ci: picongpu

psychocoderHPC committed:
fix wrong rydberg energy

In #5131 we changed rydberg energy from eV to joule but missed to apply
the conversion in the code.

psychocoderHPC committed:
fix MaxwellJuettner functor

The PR #5133 and #5142 got merged together with #4826.
Removed files and renamings for the unit systems collide therefore.

psychocoderHPC commented:
I have set this PR to draft. IT shows always green but is not ready to be merged. 

Never the less we added many changes to the dev branch therefore a rebase is required.

fabidie committed:
remove Traits.hpp

fabidie committed:
resolve rebasing errors

fabidie committed:
resolve rebasing errors

fabidie committed:
resolve rebasing errors

fabidie committed:
resolve rebasing errors

fabidie committed:
resolve rebasing errors

fabidie committed:
fixing rebasing errors

fabidie commented:
I closed this one because there happened too much the last 14 days and rebasing is a mess. There is a new one though.
